src/prometheus_metrics.rs

```diff
@@ -60,63 +68,77 @@ pub fn serve_prometheus(config: Option<&PrometheusConfig>, lemmy_context: LemmyC
 }
 
 // handler for the /metrics path
-async fn metrics(context: web::Data<Arc<PromContext>>) -> impl Responder {
-  // collect metrics
-  collect_db_pool_metrics(&context).await;
-
+async fn metrics(_context: web::Data<Arc<PromContext>>) -> impl Responder {
   let mut buffer = Vec::new();
   let encoder = TextEncoder::new();
 
-  // gather metrics from registry and encode in prometheus format
+  // gather metrics from registry
   let metric_families = prometheus::gather();
-  encoder.encode(&metric_families, &mut buffer).unwrap();
-  let output = String::from_utf8(buffer).unwrap();
 
-  HttpResponse::Ok().body(output)
+  // encode in prometheus format
+  if let Err(err) = encoder.encode(&metric_families, &mut buffer) {
+    eprintln!("Prometheus encoding error: {}", err);
+    return HttpResponse::InternalServerError().finish();
+  }
+
+  // convert to utf-8 and return
+  match String::from_utf8(buffer) {
+    Ok(body) => HttpResponse::Ok().body(body),
+    Err(err) => {
+      eprintln!("Prometheus utf-8 encoding error: {}", err);
+      HttpResponse::InternalServerError().finish()
+    }
+  }
 }
 
-// create lemmy_db_pool_* metrics and register them with the default registry
-fn create_db_pool_metrics() -> DbPoolMetrics {
-  let metrics = DbPoolMetrics {
-    max_size: Gauge::with_opts(Opts::new(
-      "lemmy_db_pool_max_connections",
-      "Maximum number of connections in the pool",
-    ))
-    .unwrap(),
-    size: Gauge::with_opts(Opts::new(
+impl DbPoolMetrics {
+  fn new(context: Arc<PromContext>) -> Self {
+    let size = Gauge::with_opts(Opts::new(
       "lemmy_db_pool_connections",
       "Current number of connections in the pool",
     ))
-    .unwrap(),
-    available: Gauge::with_opts(Opts::new(
+    .expect("Prometheus DbPoolMetrics: failed to create size metric");
+
+    let max_size = Gauge::with_opts(Opts::new(
+      "lemmy_db_pool_max_connections",
+      "Maximum number of connections in the pool",
+    ))
+    .expect("Prometheus DbPoolMetrics: failed to create max_size metric");
+
+    let available = Gauge::with_opts(Opts::new(
       "lemmy_db_pool_available_connections",
       "Number of available connections in the pool",
     ))
-    .unwrap(),
-  };
+    .expect("Prometheus DbPoolMetrics: failed to create available metric");
```

Comment:
Doesnt hurt to use `?` operator and return result here (unless that doesnt compile for some reason).